line_profiler/_line_profiler.pyx

```diff
@@ -1,12 +1,50 @@
-from .python25 cimport PyFrameObject, PyObject, PyStringObject
+from python25 cimport PyFrameObject, PyObject, PyStringObject
+from sys import byteorder
+cimport cython
+from cpython.version cimport PY_VERSION_HEX
+from libc.stdint cimport int64_t
 
+from libcpp.unordered_map cimport unordered_map
+
+# long long int is at least 64 bytes assuming c99
+ctypedef unsigned long long int uint64
+ctypedef long long int int64
 
 cdef extern from "frameobject.h":
+    """
+    inline PyObject* get_frame_code(PyFrameObject* frame) {
+        #if PY_VERSION_HEX < 0x030B0000
+            Py_INCREF(frame->f_code->co_code);
+            return frame->f_code->co_code;
+        #else
+            PyCodeObject* code = PyFrame_GetCode(frame);
+            PyObject* ret = PyCode_GetCode(code);
+            Py_DECREF(code);
+            return ret;
+        #endif
+    }
+    """
+    cdef object get_frame_code(PyFrameObject* frame)
     ctypedef int (*Py_tracefunc)(object self, PyFrameObject *py_frame, int what, PyObject *arg)
 
 cdef extern from "Python.h":
+    """
+    
+    #if PY_VERSION_HEX >= 0x030b00a6
+      #ifndef Py_BUILD_CORE
+        #define Py_BUILD_CORE 1
+      #endif
+      #include "internal/pycore_frame.h"
+      #include "cpython/code.h"
+      #include "pyframe.h"
+    #endif
+    """
+    
+    ctypedef struct PyFrameObject
+    ctypedef struct PyCodeObject
     ctypedef long long PY_LONG_LONG
     cdef bint PyCFunction_Check(object obj)
+    cdef int PyCode_Addr2Line(PyCodeObject *co, int byte_offset)
```

Comment:
Would it be better to get this declaration by simply including the header from `cpython/code.h`?

Reply:
Even if the header is included (which it is on line 38), I believe that line is necessary so Cython knows it exists when I call it later down.

Reply:
I think I need a better understanding of how .h and .c files interplay to construct the output binaries. From what I understand a "#include" is a literal copy / paste and then all data and functions are encoded into binary, which might not include all of the functions it will call, in which case a static or dynamic linker is expected to point to other binary files with the relevant function bodies they want to define, but the original binary does need a header for any external linked function it wants to call.

Does Cython not know that you included a .h file? Or in other words, whenever you `#include` a .h file you have to write the Cython header bindings for it?

Reply:
So basically, whenever you include a .h file, you have to write bindings for it. I don't know the technical reasons as to why, but I have been unable to find a way to include .h files without bindings.

This resource may help: https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html?highlight=external#including-verbatim-c-code